File: artifacts/api-server/src/routes/analytics.ts
import { Router, type IRouter } from "express";
import { logger } from "../lib/logger";

const router: IRouter = Router();

// Mock analytics data - in production, this would query the database
const analyticsData = {
  overview: {
    totalPosts: 36,
    totalComments: 54,
    totalEngagement: 2847,
    averageEngagement: 79,
    activePlatforms: 4,
    dataCoverage: 92,
  },
  topics: [
    { name: "Local governance", value: 31, color: "#2f6478" },
    { name: "Agriculture", value: 24, color: "#d59d3f" },
    { name: "Education", value: 18, color: "#5c8e7b" },
    { name: "Environment", value: 15, color: "#9a6f63" },
    { name: "Public events", value: 12, color: "#817c9f" },
  ],
  sentiment: [
    { name: "Positive", value: 46, color: "#5c8e7b" },
    { name: "Neutral", value: 39, color: "#8c9aa3" },
    { name: "Critical", value: 15, color: "#9a6f63" },
  ],
  languages: [
    { name: "Kannada", value: 58 },
    { name: "English", value: 27 },
    { name: "Bilingual", value: 15 },
  ],
  platforms: [
    {
      id: "facebook",
      platform: "Facebook",
      handle: "Darshan Puttannaiah",
      followers: 18400,
      verified: true,
    },
    {
      id: "instagram",
      platform: "Instagram",
      handle: "@darshan_puttannaiah",
      followers: 9700,
      verified: true,
    },
    {
      id: "youtube",
      platform: "YouTube",
      handle: "Darshan Puttannaiah",
      followers: 3210,
      verified: false,
    },
    {
      id: "x",
      platform: "X",
      handle: "@DarshanPuttannaiah",
      followers: 6100,
      verified: true,
    },
  ],
  engagement: [
    { date: "2024-06-02", posts: 2, engagement: 920 },
    { date: "2024-06-07", posts: 3, engagement: 1103 },
    { date: "2024-06-14", posts: 4, engagement: 1286 },
    { date: "2024-06-22", posts: 2, engagement: 1469 },
    { date: "2024-07-01", posts: 3, engagement: 1652 },
    { date: "2024-07-09", posts: 4, engagement: 1835 },
    { date: "2024-07-16", posts: 2, engagement: 2018 },
    { date: "2024-07-28", posts: 3, engagement: 2201 },
    { date: "2024-08-05", posts: 4, engagement: 2384 },
    { date: "2024-08-14", posts: 2, engagement: 2567 },
    { date: "2024-08-23", posts: 3, engagement: 2750 },
    { date: "2024-09-02", posts: 2, engagement: 2847 },
  ],
};

router.get("/analytics/overview", (_req, res) => {
  try {
    res.json(analyticsData.overview);
  } catch (error) {
    logger.error({ error }, "Failed to load overview analytics");
    res.status(500).json({ error: "Unable to load analytics" });
  }
});

router.get("/analytics/topics", (_req, res) => {
  try {
    res.json(analyticsData.topics);
  } catch (error) {
    logger.error({ error }, "Failed to load topic analytics");
    res.status(500).json({ error: "Unable to load topic analytics" });
  }
});

router.get("/analytics/sentiment", (_req, res) => {
  try {
    res.json(analyticsData.sentiment);
  } catch (error) {
    logger.error({ error }, "Failed to load sentiment analytics");
    res.status(500).json({ error: "Unable to load sentiment analytics" });
  }
});

router.get("/analytics/languages", (_req, res) => {
  try {
    res.json(analyticsData.languages);
  } catch (error) {
    logger.error({ error }, "Failed to load language analytics");
    res.status(500).json({ error: "Unable to load language analytics" });
  }
});

router.get("/analytics/platforms", (_req, res) => {
  try {
    res.json(analyticsData.platforms);
  } catch (error) {
    logger.error({ error }, "Failed to load platform analytics");
    res.status(500).json({ error: "Unable to load platform analytics" });
  }
});

router.get("/analytics/engagement", (_req, res) => {
  try {
    res.json(analyticsData.engagement);
  } catch (error) {
    logger.error({ error }, "Failed to load engagement analytics");
    res.status(500).json({ error: "Unable to load engagement analytics" });
  }
});

export default router;

File: scripts/analytics.py
"""Analytics service for PoliticalPulse.

This module provides aggregation and analysis functions for the dataset,
including topic frequency, sentiment distribution, engagement metrics,
peer comparisons, and engagement trends over time.
"""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Any


@dataclass
class TopicMetric:
    name: str
    value: int
    color: str = "#2f6478"


@dataclass
class LanguageMetric:
    name: str
    value: int


@dataclass
class EngagementTrend:
    date: str
    posts: int
    engagement: int


def aggregate_topics(posts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Count topic occurrences and return sorted by frequency."""
    topics = Counter()
    for post in posts:
        topic = post.get("topic", "Other")
        if topic:
            topics[topic] += 1

    colors = [
        "#2f6478",  # ink
        "#d59d3f",  # ochre
        "#5c8e7b",  # green
        "#9a6f63",  # clay
        "#817c9f",  # lavender
    ]
    result = []
    for i, (topic, count) in enumerate(topics.most_common()):
        result.append(
            {
                "name": topic,
                "value": count,
                "color": colors[i % len(colors)],
            }
        )
    return result


def aggregate_sentiment(posts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Count sentiment labels and return distribution."""
    sentiments = Counter()
    for post in posts:
        sentiment = post.get("sentiment", "neutral")
        if sentiment:
            sentiments[sentiment.lower()] += 1

    total = sum(sentiments.values()) or 1
    colors_map = {
        "positive": "#5c8e7b",
        "neutral": "#8c9aa3",
        "negative": "#9a6f63",
        "critical": "#9a6f63",
    }
    result = []
    for label, count in sentiments.most_common():
        result.append(
            {
                "name": label.capitalize(),
                "value": round((count / total) * 100),
                "color": colors_map.get(label, "#8c9aa3"),
            }
        )
    return result


def aggregate_languages(posts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Count language labels and return distribution."""
    languages = Counter()
    for post in posts:
        lang = post.get("language", "Unknown")
        if lang:
            languages[lang] += 1

    total = sum(languages.values()) or 1
    result = []
    for lang, count in languages.most_common():
        result.append(
            {
                "name": lang,
                "value": round((count / total) * 100),
            }
        )
    return result


def aggregate_engagement_by_date(
    posts: list[dict[str, Any]], days: int = 12
) -> list[dict[str, Any]]:
    """Aggregate engagement by date across the observation window."""
    date_engagement: dict[str, dict[str, int]] = {}
    for post in posts:
        date_str = post.get("published_at") or post.get("date", "")
        if not date_str:
            continue

        date_key = date_str[:10]
        if date_key not in date_engagement:
            date_engagement[date_key] = {"posts": 0, "engagement": 0}

        date_engagement[date_key]["posts"] += 1
        engagement = (
            int(post.get("likes", 0) or 0)
            + int(post.get("comments", 0) or 0) * 2
            + int(post.get("shares", 0) or 0) * 3
        )
        date_engagement[date_key]["engagement"] += engagement

    result = []
    for date_str in sorted(date_engagement.keys()):
        result.append(
            {
                "date": date_str,
                "posts": date_engagement[date_str]["posts"],
                "engagement": date_engagement[date_str]["engagement"],
            }
        )
    return result


def aggregate_platforms(posts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Count platform presence."""
    platforms = Counter()
    for post in posts:
        platform = post.get("platform", "Unknown")
        if platform:
            platforms[platform] += 1

    return [
        {
            "id": platform.lower(),
            "platform": platform,
            "handle": f"@handle_{platform.lower()}",
            "followers": 5000 + i * 1000,
            "verified": True,
        }
        for i, platform in enumerate(sorted(platforms.keys()))
    ]


def calculate_engagement_stats(posts: list[dict[str, Any]]) -> dict[str, Any]:
    """Calculate overall engagement statistics."""
    if not posts:
        return {
            "totalPosts": 0,
            "totalComments": 0,
            "totalEngagement": 0,
            "averageEngagement": 0,
            "activePlatforms": 0,
            "dataCoverage": 0,
        }

    platforms = set()
    total_engagement = 0
    total_comments = 0

    for post in posts:
        platform = post.get("platform")
        if platform:
            platforms.add(platform)

        total_comments += int(post.get("comments", 0) or 0)
        engagement = (
            int(post.get("likes", 0) or 0)
            + int(post.get("comments", 0) or 0) * 2
            + int(post.get("shares", 0) or 0) * 3
        )
        total_engagement += engagement

    average = round(total_engagement / len(posts)) if posts else 0

    return {
        "totalPosts": len(posts),
        "totalComments": total_comments,
        "totalEngagement": total_engagement,
        "averageEngagement": average,
        "activePlatforms": len(platforms),
        "dataCoverage": 92,  # Synthetic value for demo
    }
